### PR TITLE
feat: add priorityclassname and lifecycle

### PR DESCRIPTION
**Pull request by rtest12**
### Summary

Task: [USAWS-56](https://saritasa.atlassian.net/browse/USAWS-56)

- [add priorityClassName for the deployment](https://github.com/stakater/application/issues/298)
- [add lifeCycle block to the deployment container](https://github.com/stakater/application/issues/299)
- Fork their repo, make changes in our saritasa-nest org and request PR
- Switch our addons in usummit to use saritasa-nest repo for the application
- Add preStop condition to all apps we currently have

Screenshot from 2024-05-29 17-55-36
Screenshot from 2024-05-28 22-04-47

[As logs for a Hook handler are not exposed in Pod events](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#debugging-hook-handlers), here's proof that it works

Screenshot from 2024-05-29 17-43-03
Screenshot from 2024-05-29 17-42-14

Screenshot from 2024-06-03 14-55-37

[USAWS-56]: https://saritasa.atlassian.net/browse/USAWS-56?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: application/templates/cronjob.yaml [hunk 1]
```diff
@@ -100,6 +98,9 @@ spec:
           {{- with $job.affinity }}
           affinity:
 {{ toYaml . | indent 12 }}
+          {{- end }}
+          {{- with $job.priorityClassName }}
+          priorityClassName: {{ toYaml . | nindent 12 }}
```
Comment (kseniyashaydurova) [inline application/templates/cronjob.yaml]: Did you test that it really works? Now it looks confusing, cause there would be just a single string value and there is no need in `toYaml` rendering with nindent

File: application/templates/deployment.yaml [hunk 2]
```diff
@@ -75,19 +75,23 @@ spec:
...
 {{ include "application.tplvalues.render" ( dict "value" $value "context" $ ) | indent 8 }}
       {{- end }}
       {{- end }}
       {{- if .Values.deployment.nodeSelector }}
       nodeSelector:
 {{ toYaml .Values.deployment.nodeSelector | indent 8 }}
+      {{- end }}
+      {{- if .Values.deployment.priorityClassName }}
+      priorityClassName:
+{{ toYaml .Values.deployment.priorityClassName | indent 8 }}
```
Comment (kseniyashaydurova) [inline application/templates/deployment.yaml]: The same as above

File: application/templates/deployment.yaml [hunk 3]
```diff
@@ -303,4 +307,12 @@ spec:
       hostNetwork: {{ .Values.deployment.hostNetwork }}
       {{- end }}
       terminationGracePeriodSeconds: {{ .Values.deployment.terminationGracePeriodSeconds }}
+      {{- if .Values.lifecycle }}
+      lifecycle:
+        {{- if .Values.lifecycle.preStop }}
```
Comment (kseniyashaydurova) [inline application/templates/deployment.yaml]: Let's just add here like this one:

```
lifecycle:
  {{ toYaml .Values.deployment.lifecycle | indent 8 }}
```

No need in separate `preStop` definition

File: application/templates/deployment.yaml [hunk 4]
```diff
@@ -303,4 +307,12 @@ spec:
       hostNetwork: {{ .Values.deployment.hostNetwork }}
       {{- end }}
       terminationGracePeriodSeconds: {{ .Values.deployment.terminationGracePeriodSeconds }}
+      {{- if .Values.lifecycle }}
```
Comment (kseniyashaydurova) [inline application/templates/deployment.yaml]:
```suggestion
      {{- if .Values.deployment.lifecycle }}
```

File: application/values-test.yaml [hunk 5]
```diff
@@ -223,6 +225,12 @@ deployment:
   # Graceful termination timeout
   terminationGracePeriodSeconds: 120
 
+  # Container lifecycle management for graceful startup and shutdown procedures
+  lifecycle:
```
Comment (kseniyashaydurova) [inline application/values-test.yaml:229]: Why it is not inside `deployment`?

File: application/values.yaml [hunk 6]
```diff
@@ -291,6 +292,13 @@ deployment:
   # Graceful termination timeout
   terminationGracePeriodSeconds:
 
+  # Default lifecycle configuration
+  lifecycle: {}
```
Comment (kseniyashaydurova) [inline application/values.yaml:325]: Why it is not ionside `deployment`?

**Review by kseniyashaydurova — CHANGES_REQUESTED** [review]
Pls:
1. Provide proofs that `values-test.yaml` really works
2. Update any usummit component to use our custom stakater with priorityclass and lifecycle and add here link to it (or proffs that everything really works)

File: application/templates/cronjob.yaml (hunk 1 above)
Comment (rtest12) [reply application/templates/cronjob.yaml]: @kseniyashaydurova 
Fixed

File: application/templates/deployment.yaml (hunk 2 above)
Comment (rtest12) [reply application/templates/deployment.yaml]: @kseniyashaydurova 
Fixed

File: application/templates/deployment.yaml (hunk 3 above)
Comment (rtest12) [reply application/templates/deployment.yaml]: @kseniyashaydurova 
Done

File: application/values-test.yaml (hunk 5 above)
Comment (rtest12) [reply application/values-test.yaml:229]: @kseniyashaydurova 
It's in deployment
I think you were misled by the huge number of formatting errors in the source repository - vscode just went crazy and didn't show the nesting of variables correctly.  I've fixed it now and it's displaying correctly.
Screenshot from 2024-05-28 14-37-31

File: application/values.yaml (hunk 6 above)
Comment (rtest12) [reply application/values.yaml:325]: @kseniyashaydurova 
The same

**Comment by rtest12** [comment]
> Pls:
> 
>     1. Provide proofs that `values-test.yaml` really works
> 
>     2. Update any usummit component to use our custom stakater with priorityclass and lifecycle and add here link to it (or proffs that everything really works)

@kseniyashaydurova 
Ptl that the lifecycle location is correct, I'm not sure
screenshots are attached to PR

File: application/values-test.yaml [hunk 7]
```diff
@@ -1286,22 +1293,22 @@ grafanaDashboard:
         matchLabels:
           app: test-2
 
-cronJob: 
+cronJob:
```
Comment (kseniyashaydurova) [inline application/values-test.yaml:1296]: Add `priorityClass` for Cronjob for the test

File: application/values.yaml [hunk 8]
```diff
@@ -14,31 +14,31 @@ applicationName: "application"
...
-# These labels will be added on all resources, 
-# and you can add additional labels from below 
+# These labels will be added on all resources,
+# and you can add additional labels from below
 # on individual resource
 ##########################################################
-  
-cronJob: 
+
+cronJob:
```
Comment (kseniyashaydurova) [inline application/values.yaml:22]: Add `priorityClass` to cronjob too (in the commented `db-migration` block)

File: application/values.yaml [hunk 9]
```diff
@@ -148,6 +148,9 @@ deployment:
     # volume-name-2:
     #    mountPath: path-2
 
+  # priorityClassName defines the priority class for pod scheduling
+  priorityClassName:
```
Comment (kseniyashaydurova) [inline application/values.yaml]:
```suggestion
  priorityClassName: ""
```

**Review by kseniyashaydurova — CHANGES_REQUESTED** [review]
Pls also add proofs that values_test.yaml correctly works, i.e. generate manifests with `values_test.yaml` and show that there are no errors and `priorityClassName` were correctly added to deployment and cronjob and the same for `lifecycle` (you will need that on opening PR to stakater to prove them that your changes are fine and won't break anything)

File: application/values.yaml (hunk 8 above)
Comment (rtest12) [reply application/values.yaml:22]: @kseniyashaydurova 
Done

File: application/values-test.yaml (hunk 7 above)
Comment (rtest12) [reply application/values-test.yaml:1296]: @kseniyashaydurova
Done

**Review by kseniyashaydurova — APPROVED** [review]

**Review by KrawczowaKris — APPROVED** [review]

**Review by dmitry-mightydevops — APPROVED** [review]
@rtest12 @kseniyashaydurova 
- pls open PR into the upstream repo
- when approved - pls update all our apps/addons to use newer version
- pls setup priorityClassName for all apps (to be high-priority class)
- pls setup lifecycle for frontend type pods

# In the future
- one change == one PR (you have 2 changes here) so will be harder to approve the PR with the upstream

**Comment by rtest12** [comment]
PR in the original repository was merged, i'll close this PR as unnecessary
https://github.com/stakater/application/pull/313#event-13518077134

**Pull request closed without merging**